### PR TITLE
[compiler-v2] Receiver functions

### PR DESCRIPTION
**Pull request by wrwg**
This is a complete implementation of receiver functions following the language design as outline in the "Future of Move" blog.

- Receiver functions are declared using the well-known parameter name `self` as the first parameter.
- A receiver function can be either called as `x.f()` or `f(x)` (for downwards compatibility)
- Only structs and vector types can be receiver arguments.
- If a receiver parameter has type `&T` or `&mut T`, the argument will be automatically borrowed as needed.
- Receiver functions can only be declared for types in the same module. This is not a technical restriction but a methodological, to prevent confusing code adding receivers to arbitrary framework types.
- A wrongly declared receiver function currently only leads to a warning. This is for downwards compatibility (also see #12219).
- Partial instantiated generic receiver functions (as in `fun f(self: T<u64>)` are currently not allowed (but also see #12221).
- Receiver functions are fully integrated into type inference. Hence for `x.f()`, if the type of `x` is not known at checking time, the resolution `f` will be deferred until it is. This is technically handled via a new type constraint `SomeReceiverFunction` and the finalization of receiver calls in `ExpTranslator::post_process_body`.

This adds a number of tests in `checking/receiver`.

### CODE REVIEW

**Review by brmataptos — COMMENTED** [review]
I don't see a test case with multiple `receiver` methods available.  Do we do any overload resolution here?  Can we select a `receiver` from another module?

**Comment by wrwg** [comment]
> I don't see a test case with multiple `receiver` methods available. Do we do any overload resolution here? Can we select a `receiver` from another module?

Added one in `same_names.move`.

We do not exactly have overloading right now but the same receiver function name can be used for different types. Yet those receiver functions need to be declared in different modules since we do not support overloading inside a single module. We may later revisit and relax this.

File: third_party/move/move-compiler-v2/tests/checking/receiver/call_errors.exp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+
+Diagnostics:
+error: undeclared receiver function `receiver_wrongly_spelled` for type `S<u64>`
+   ┌─ tests/checking/receiver/call_errors.move:16:9
+   │
+16 │         s.receiver_wrongly_spelled(x);
+   │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: expected 2 items but found 3
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/receiver/call_errors.exp:9]: nit: while this is technically correct, we may want specialize an error message for receiver style calls (something like: expected 2 arguments for the receiver function, but found 1). On a quick parse of the error message, the user may think the compiler is wrong.

File: third_party/move/move-compiler-v2/tests/checking/receiver/generic_calls.move [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    fun receiver_more_generics<T, R>(self: S<T>, _y: R) {
+    }
+
+    fun test_call_styles(s: S<u64>, x: u64) {
+        s.receiver(x);
+        s.receiver_ref(x);
+        s.receiver_ref_mut(x);
+        s.receiver_more_generics(22)
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/receiver/generic_calls.move:25]: 1. Could we please add a "passing" case for the turbofish operator (`::<>`)?
2. Could we please add a case with `inline` receiver functions?

File: third_party/move/move-compiler-v2/tests/checking/receiver/vectors.move [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Do not include stdlib for this test:
+// no-stdlib
+
+module 0x1::vector { // must be this module
+
+    fun receiver<T>(self: vector<T>, _y: T) {
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/receiver/vectors.move:6]: There are currently several methods in the standard vector module which would benefit from receiver syntax (eg, `reverse`). Is the plan to eventually rewrite the standard modules to use `self` where possible (which should be upgrade compatible due to change in only param names, assuming in all applicable cases they were the initial argument)?

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 4]
```diff
@@ -1736,7 +1734,29 @@ fn parse_dot_or_index_chain(context: &mut Context) -> Result<Exp, Box<Diagnostic
             Tok::Period => {
                 context.tokens.advance()?;
                 let n = parse_identifier(context)?;
-                Exp_::Dot(Box::new(lhs), n)
+                let ahead = context.tokens.peek();
+                if matches!(ahead, Tok::LParen | Tok::ColonColon) {
+                    let generics = if ahead == Tok::ColonColon {
```
Comment (vineethk) [inline third_party/move/move-compiler/src/parser/syntax.rs:1739]: I am assuming this turbofish operator `::<>` is implemented to allow for future modifications to the language? Or is there a case in the current language where we cannot type infer and need the user to explicitly specify the receiver function instantiation?

**Review by vineethk — COMMENTED** [review]
Looking good so far. Nice addition to the language!

**Review by brmataptos — COMMENTED** [review]
Still reading, but a few comments.

**Review by wrwg — COMMENTED** [review]
Thanks for the helpful reviews, PTAL.

Comments are addressed unless I overlooked some. This PR also enables instantiation ability checks which were delayed for landing this, however, to avoid overloading this PR, I haven't followed up on fully testing this, and also not removed now obsolete instantiation checks in ability_processor. This should happen in another PR.

**Review by vineethk — APPROVED** [review]
LGTM

**Comment by wrwg** [comment]
Notice that their are some failures in ability checking in move-to-yul and perhaps elsewhere, they will be fixed in the next PR.

**Review by brmataptos — COMMENTED** [review]
Still reading, but this seems worth communicating.

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 5]
```diff
@@ -1681,6 +1760,131 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
...
+                                    // and it's kind.
+                                    let borrow_kind_opt =
+                                        inst.receiver_needs_borrow(&receiver_arg_ty);
+                                    if !inst.type_inst.is_empty() {
+                                        // We need to annotate the instantiation of the function
+                                        // at the node. To obtain it, unification needs to be run
+                                        // again. If unification fails, errors will have been
+                                        // already reported, so we can ignore the result.
+                                        let mut subs = self.subs.clone();
+                                        let mut ok = true;
```
Comment (brmataptos) [inline third_party/move/move-model/src/builder/exp_builder.rs:1796]: This code is pretty unreadable.  It seems like it would be so much better to put it in a helper function (or even in a new method of `Substution` so code can be clarified by early returns and the `?` operator, e.g.:
```
let unify_instantiation = |subs|: Result<Subs, TypeUnificationError> = {
    if let Some(tys) = generics {
        subs.unify_vec_maybe_type_args(...)?;
    }
    if let Some(ref_kind) = &borrow_kind_opt {
        ...
    }
    subs.unify_vec(...)?;
    subs.unify(...)?;
    OK(subs)
}
if let Ok(new_subs) = unify_instantiation(self.subs.clone()) {
    self.subs = new_subs;
    ...
```
or something similar.

**Review by brmataptos — COMMENTED** [review]
Looks mostly ok.  Will approve if you can add a test requiring receiver freeze and it works.

File: third_party/move/move-compiler-v2/tests/checking/receiver/same_names.move [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+module 0x42::c {
+    use 0x42::a;
+    use 0x42::b;
+
+    inline fun foo(f: |a::MyList, b::MyOtherList|, x: a::MyList, y: b::MyOtherList) {
+        f(x, y)
+    }
+
+    fun test(x: a::MyList, y: b::MyOtherList) {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/receiver/same_names.move:27]: It does not work for inline functions. This is all type checking and completely unrelated to them. There is no interleaving, not sure why you say this, inlining strictly runs after type checking has been 100% done.

It just doesn't work because Move is not C++. If you have `fun f<X>(x: X) { ... }` there is nothing known about `x` and which functions you can call on it.

File: third_party/move/move-compiler-v2/tests/checking/typing/assign_wrong_arity.exp [hunk 7]
```diff
@@ -1,6 +1,6 @@
 
 Diagnostics:
-error: the left-hand side has 3 items but the right-hand side provided 0
+error: the left-hand side has 0 items but the right-hand side provided 3
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/typing/assign_wrong_arity.exp:3]: Please go to `ty.rs`, it is there since a few weeks, and a central place where you find all those messages, and improve them. It is easy.

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 5 above)
Comment (wrwg) [reply third_party/move/move-model/src/builder/exp_builder.rs]: I pulled this code into its own method and refactored it a bit.

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 8]
```diff
@@ -1681,6 +1760,131 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
...
+                                                        true,
+                                                        Variance::NoVariance,
+                                                        WideningOrder::LeftToRight,
+                                                        None,
+                                                        &inst.type_inst,
+                                                        &tys,
+                                                    )
+                                                    .is_ok()
+                                        }
+                                        if let Some(ref_kind) = &borrow_kind_opt {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/builder/exp_builder.rs]: This is now done differently, all those checks should be done whether ok or not.

File: third_party/move/evm/move-to-yul/tests/GlobalVectors.exp
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/evm/move-to-yul/tests/GlobalVectors.exp:1]: Added a comment to the PR already before -- ability checking will be fixed when the next PR lands, which is out for review. This doesn't effect non-test code so we can move forward.

File: third_party/move/move-compiler-v2/tests/checking/receiver/calls.move [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    fun receiver_ref_mut(self: &mut S, y: u64): u64 {
+        self.x + y
+    }
+
+    inline fun inline_receiver_ref_mut(self: &mut S, y: u64): u64 {
+        self.x + y
+    }
+
+    fun test_call_styles(s: S): u64 {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/receiver/calls.move:23]: Good catch. Added your test and fixed freezing, which was added after this PR was created.

**Review by wrwg — COMMENTED** [review]
Thanks for the review

File: third_party/move/move-compiler-v2/tests/checking/receiver/calls.move (hunk 9 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/receiver/calls.move:23]: PS. Did not add a transactional test because really nothing is different anywhere after the model builder.

**Comment by wrwg** [comment]
Added now a transactional test with receiver and freeze.

**Review by brmataptos — APPROVED** [review]

File: third_party/move/evm/move-to-yul/tests/GlobalVectors.exp
Comment (brmataptos) [inline third_party/move/evm/move-to-yul/tests/GlobalVectors.exp:1]: You still have this unrelated file.  It would be nice to fix it in this PR, rather than another PR, where reviewers have to go to great pains to see that the code went back to what it was supposed to be.

**Review by brmataptos — COMMENTED** [review]
What's the deal with `third_party/move/evm/move-to-yul/tests/GlobalVectors.exp`?  Why would that change here?  I see you change it back in your next PR.  It may be easier to fix it here (I assume it's needed for the CI tests to pass) since the changes are impossible to understand.

File: third_party/move/evm/move-to-yul/tests/GlobalVectors.exp
Comment (wrwg) [reply third_party/move/evm/move-to-yul/tests/GlobalVectors.exp:1]: I answered this already before and I even send a message before you even started reviewing:

>  Notice that their are some failures in ability checking in move-to-yul and perhaps elsewhere, they will be fixed in the next PR.

In fact they are fixed #12675
